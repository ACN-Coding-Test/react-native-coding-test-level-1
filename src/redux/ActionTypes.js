export default {
    SET_ITEMS: "SET_ITEMS",
    VIEW_ITEM: "VIEW_ITEM"
}
