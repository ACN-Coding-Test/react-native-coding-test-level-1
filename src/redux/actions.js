import { ToastAndroid } from "react-native";
import { UPDATE_POKEMON_LIST, UPDATE_POKEMON_DATA } from "./types";
import { getService } from "../services/services";
import * as STRINGS from "../utils/strings";

export const updatePokemonList = (url) => (dispatch) => {
    return new Promise((resolve, reject) => {
        getService(url)
            .then(response => {
                dispatch({
                    type: UPDATE_POKEMON_LIST,
                    payload: response
                })
                return resolve(response);
            }).catch(error => {
                ToastAndroid.show(STRINGS.LIST_FETCH_ERROR, ToastAndroid.BOTTOM, ToastAndroid.CENTER);
                return reject(error)
            })
    })
}

export const updatePokemonData = (url) => (dispatch) => {
    return new Promise((resolve, reject) => {
        getService(url)
            .then(response => {
                dispatch({
                    type: UPDATE_POKEMON_DATA,
                    payload: response
                })
                return resolve(response);
            }).catch(error => {
                ToastAndroid.show(STRINGS.LIST_FETCH_ERROR, ToastAndroid.BOTTOM, ToastAndroid.CENTER);
                return reject(error)
            })
    })
}