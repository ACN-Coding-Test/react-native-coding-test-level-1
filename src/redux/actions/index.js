import pokemonActions from './pokemon';

const allActions = {
  pokemonActions
};

export default allActions;
