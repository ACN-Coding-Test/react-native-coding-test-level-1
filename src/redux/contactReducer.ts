import {createSlice, createReducer, PayloadAction} from '@reduxjs/toolkit';
import {AppThunk} from './index';

export type Info = {
    name: string;
    email: string;
};

type ContactType = {
    loading: boolean;
    info: Info;
    date: string;
};

const contactState = {
    loading: false,
    info: {
        name: '',
        email: ''
    },
    date: new Date().toDateString()
} as ContactType;

const contactSlice = createSlice({
    name: 'contact',
    initialState: contactState,
    reducers: {
        setLoadingStart: state => {
            state.loading = true;
        },

        setLoadingEnd: state => {
            state.loading = false;
        },

        onChangeInfo: (
            state,
            {
                payload
            }: PayloadAction<{
                key: keyof Info;
                value: string;
            }>
        ) => {
            let {key, value} = payload;
            state.info[key] = value;
        },

        onChangeBOD: (
            state,
            {
                payload
            }: PayloadAction<{
                value: string;
            }>
        ) => {
            let {value} = payload;

            state.date = value;
        }
    }
});

export const {setLoadingStart, setLoadingEnd, onChangeInfo, onChangeBOD} =
    contactSlice.actions;

export default contactSlice.reducer;
