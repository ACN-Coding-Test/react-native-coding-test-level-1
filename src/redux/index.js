import { combineReducers, createStore, applyMiddleware } from 'redux'
import thunk from 'redux-thunk'
import item from './item/reducer'
import logger from 'redux-logger'
// Combine all
const appReducer = combineReducers({
    item
})

const middlewares = [
  thunk
]

if (__DEV__) {
  middlewares.push(logger)
}
export default createStore(appReducer, applyMiddleware(...middlewares))
