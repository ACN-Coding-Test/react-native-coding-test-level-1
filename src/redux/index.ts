import {configureStore, Action} from '@reduxjs/toolkit';
import {combineReducers} from 'redux';
import {ThunkAction} from 'redux-thunk';

import ContactReducer from './contactReducer';
import PokemonReducer from './pokemonReducer';

const reducer = combineReducers({
    contact: ContactReducer,
    pokemon: PokemonReducer
});

export const store = configureStore({
    reducer: reducer
});

export type RootState = ReturnType<typeof reducer>;

export type AppDispatch = typeof store.dispatch;

export type AppThunk = ThunkAction<void, RootState, unknown, Action<string>>;
