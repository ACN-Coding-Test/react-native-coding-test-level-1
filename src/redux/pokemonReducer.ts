import {createSlice, createReducer, PayloadAction} from '@reduxjs/toolkit';
import {AppThunk} from './index';
import {Dispatch} from 'redux';

export type PokemonList = {
  name: string;
  url: string;
};

type PokemonListResponse = {
  count: number;
  next: string | null;
  previous: string | null;
  results: Array<PokemonList>;
};

export type PokemonInfo = {
  id: number;
  name: string;
  height: number;
  weight: number;
  types: Array<{
    slot: number;
    type: {
      name: string;
      url: string;
    };
  }>;
  abilities: Array<{
    ability: {
      name: string;
      url: string;
    };
    is_hidden: false;
    slot: number;
  }>;
  base_experience: number;
  forms: Array<{
    name: string;
    url: string;
  }>;
  game_indices: Array<{
    game_index: number;
    version: {
      name: string;
      url: string;
    };
  }>;

  location_area_encounters: string;
  moves: Array<{
    move: {
      name: string;
      url: string;
    };
    version_group_details: Array<{
      level_learned_at: number;
      move_learn_method: {
        name: string;
        url: string;
      };
      version_group: {
        name: string;
        url: string;
      };
    }>;
  }>;
  stats: Array<{
    base_stat: number;
    effort: number;
    stat: {
      name: string;
      url: string;
    };
  }>;
};

type Pokemon = {
  loading: boolean;
  pokemon_count: number;
  list: Array<PokemonList>;
  pokemonInfo: PokemonInfo;
  offset: number;
};

const pokemonState = {
  loading: false,
  pokemon_count: 0,
  list: [],
  pokemonInfo: {
    id: 0,
    name: '',
    height: 0,
    weight: 0,
    types: [],
    abilities: [],
    base_experience: 0,
    forms: [],
    game_indices: [],
    location_area_encounters: '',
    moves: [],
    stats: [],
  },
  offset: 10,
} as Pokemon;

const pokemonSlice = createSlice({
  name: 'contact',
  initialState: pokemonState,
  reducers: {
    setLoadingStart: state => {
      state.loading = true;
    },

    setLoadingEnd: state => {
      state.loading = false;
    },

    onSetPokemonList: (
      state,
      {
        payload,
      }: PayloadAction<{
        value: Array<PokemonList>;
      }>,
    ) => {
      let {value} = payload;
      state.list = value;
    },

    onSetPokemonInfo: (
      state,
      {
        payload,
      }: PayloadAction<{
        value: PokemonInfo;
      }>,
    ) => {
      let {value} = payload;
      state.pokemonInfo = value;
    },

    addOffset: state => {
      state.offset = state.offset + 10;
    },

    deductOffset: state => {
      state.offset = state.offset - 10;
    },

    setPokemonCount: (
      state,
      {
        payload,
      }: PayloadAction<{
        value: number;
      }>,
    ) => {
      const {value} = payload;
      state.pokemon_count = value;
    },
  },
});

export const {
  setLoadingStart,
  setLoadingEnd,
  onSetPokemonInfo,
  onSetPokemonList,
  addOffset,
  deductOffset,
  setPokemonCount,
} = pokemonSlice.actions;

export const getPokemonList =
  (offset: number): AppThunk =>
  async (dispatch: Dispatch) => {
    try {
      dispatch(setLoadingStart());

      let response = await fetch(
        `https://pokeapi.co/api/v2/pokemon?offset=${offset}&limit=10`,
        {
          method: 'GET',
        },
      );

      if (response.status === 200) {
        let data: PokemonListResponse = await response.json();
        dispatch(setPokemonCount({value: data.count}));
        dispatch(onSetPokemonList({value: data.results}));
      }

      dispatch(setLoadingEnd());
    } catch (error) {
      console.log(error);

      dispatch(setLoadingEnd());
    }
  };

export const getPokemonInfo =
  (url: string): AppThunk =>
  async (dispatch: Dispatch) => {
    try {
      dispatch(setLoadingStart());

      let response = await fetch(url, {
        method: 'GET',
      });

      if (response.status === 200) {
        let data: PokemonInfo = await response.json();
        dispatch(onSetPokemonInfo({value: data}));
      }

      dispatch(setLoadingEnd());
    } catch (error) {
      console.log(error);

      dispatch(setLoadingEnd());
    }
  };

export default pokemonSlice.reducer;
