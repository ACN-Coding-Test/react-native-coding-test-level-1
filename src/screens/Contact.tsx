import React from 'react';
import {useState} from 'react';
import {View, TextInput as Input, Image, Alert, StyleSheet} from 'react-native';
import {Paragraph, Subheading, Surface, Title} from 'react-native-paper';
import {Modal} from 'native-base';
import {Button, DateInput, TextInput} from '../components';
import {Container} from '../containers';
import {HomeNavigationProp} from '../navigation/model';
import {RootState} from '../redux';
import {Info, onChangeBOD, onChangeInfo} from '../redux/contactReducer';
import {useAppDispatch, useAppSelector} from '../redux/hooks';
import {isEmailValid, isNameValid} from '../utils';
import {colors} from '../utils/themes';

const pokemonLogo = require('../../assets/pokemon_logo.png');

type Props = {
  navigation: HomeNavigationProp;
};

const Contact = () => {
  const dispatch = useAppDispatch();
  const {info, date} = useAppSelector((state: RootState) => state.contact);

  const [open, setOpen] = useState<boolean>(false);
  const validEmail = isEmailValid(info.email);
  const validName = isNameValid(info.name);

  const onChangeTextInput = (key: keyof Info) => (value: string) => {
    dispatch(onChangeInfo({key: key, value: value}));
  };

  const onChangeDate = (value: Date) => {
    dispatch(onChangeBOD({value: value.toDateString()}));
  };

  const submit = () => {
    const {name, email} = info;

    if (name.length === 0 || !validName) {
      Alert.alert('Alert', 'Please enter name correctly');
      return;
    }

    if (email.length === 0 || !validEmail) {
      Alert.alert('Alert', 'Please enter email correctly');
      return;
    }

    setOpen(!open);
  };

  const parseDate = new Date(date);

  return (
    <Container
      backgroundColor="white"
      statusBarColor="red"
      title="PokeDex"
      isMain>
      <View style={styles.container}>
        <View style={styles.image_container}>
          <Image source={pokemonLogo} style={styles.image} />
        </View>
        <Subheading style={styles.text}>Gotta catch 'em all</Subheading>

        <View style={styles.input_container}>
          <TextInput
            value={info.name}
            onChange={onChangeTextInput('name')}
            label="Name"
            errorLines="Enter your name correctly."
            affix
            affixLength="/50"
            keyInfo="name"
            error={validName}
            maxlength={50}
          />

          <TextInput
            value={info.email}
            onChange={onChangeTextInput('email')}
            label="Email"
            errorLines="Enter your email correctly."
            keyInfo="email"
            error={validEmail}
          />

          <DateInput value={parseDate} onChange={onChangeDate} />
        </View>

        <View style={styles.button_container}>
          <Button
            label="Submit"
            backgroundColor={colors.blue}
            onPress={submit}
          />
        </View>

        <Modal isOpen={open} onClose={() => setOpen(!open)} size={'xl'}>
          <Modal.Content>
            <View style={styles.modal_container}>
              <View style={styles.info_container}>
                <Title style={styles.title}>Thank you for your feedback!</Title>

                <View style={{marginTop: '5%'}}>
                  <Subheading style={styles.subheading}>
                    Name : {info.name}
                  </Subheading>

                  <Subheading style={styles.subheading}>
                    Email : {info.email}
                  </Subheading>

                  <Subheading style={styles.subheading}>
                    Birthday : {date}
                  </Subheading>
                </View>
              </View>
            </View>
          </Modal.Content>
        </Modal>
      </View>
    </Container>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    paddingTop: '5%',
    paddingBottom: '5%',
  },
  image_container: {marginTop: '10%'},
  image: {resizeMode: 'contain', height: 200, width: 300, alignSelf: 'center'},
  text: {
    color: colors.blue,
    letterSpacing: 1.5,
    fontStyle: 'italic',
    fontWeight: 'bold',
    fontSize: 25,
    textAlign: 'center',
    marginTop: 10,
  },
  button_container: {marginTop: '5%', alignItems: 'center'},
  input_container: {marginTop: '5%'},
  modal_container: {paddingTop: '7.5%', paddingBottom: '7.5%'},
  info_container: {alignItems: 'center', justifyContent: 'center'},
  title: {fontSize: 20, color: colors.blue, letterSpacing: 1.5},
  subheading: {letterSpacing: 1.5},
});

export default Contact;
