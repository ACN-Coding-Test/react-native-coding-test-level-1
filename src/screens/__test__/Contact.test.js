import 'react-native';
import React from 'react';
// import {render, fireEvent} from '../../../jest/helper';
import {NativeBaseProvider} from 'native-base';
import {NavigationContainer} from '@react-navigation/native';
import {Provider} from 'react-redux';
import Contact from '../Contact';
import {store} from '../../redux';
import renderer from 'react-test-renderer';

import {render, screen, fireEvent} from '@testing-library/react-native';

describe('form edit', () => {
  const wrapper = renderer.create(
    <Provider store={store}>
      <NativeBaseProvider>
        <Contact />
      </NativeBaseProvider>
    </Provider>,
  );

  it('Should render Button', () => {
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  test('change name input', () => {
    const component = (
      <Provider store={store}>
        <NativeBaseProvider>
          <Contact />
        </NativeBaseProvider>
      </Provider>
    );

    render(component);

    const testName = 'Ezanie';
    fireEvent.changeText(screen.getByLabelText('Name'), testName);

    expect(namefield.props.value).toEqual(testName);
  });

  test('change email input', async () => {
    const component = (
      <Provider store={store}>
        <NativeBaseProvider>
          <Contact />
        </NativeBaseProvider>
      </Provider>
    );

    render(component);

    // const emailfield = await screen.getByLabelText('emailInput');
    // expect(emailfield).toBeTruthy();
    // fireEvent.changeText(emailfield, testEmail);

    const testEmail = 'ezzadsyafik@gmail.com';
    fireEvent.changeText(screen.getByLabelText('Email'), testEmail);

    expect(emailfield.props.value).toEqual(testEmail);
  });
});
