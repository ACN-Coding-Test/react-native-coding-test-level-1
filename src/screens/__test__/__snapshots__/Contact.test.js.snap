// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`form edit Should render Button 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
/>
`;
